### PR TITLE
Rebase on upstream (take three)

### PR DESCRIPTION
**Pull request by ccasin**
Rebase of all Jane Street stuff (including comprehensions) onto the latest upstream ocamlformat.

@antalsz please look at two things in particular:

1. How I merged your tests into the upstream test refactoring I previously mentioned.
2. I moved `language_extension.{ml,mli}` into the `ocaml-common` directory rather than having copies in each parser.  It looks like they were identical and didn't depend on anything that prevents this, but let me know if there was a reason for how it was.

### CODE REVIEW

**Comment by antalsz** [comment]
Re. #2: `language_extensions.ml{,i}` comes from `ocaml-jst`.  I can't tell if `ocaml-common/` does; if it does, then that's a fine place for `language_extensions.ml{,i}` too.

File: test/passing/tests/index_op.ml [hunk 1]
```diff
@@ -169,6 +175,12 @@ let _ = Array.get [||] (-8)
 
 let _ = Array.unsafe_get [||] (-8)
 
+let _ = [::].(-8)
```
Comment (antalsz) [inline test/passing/tests/index_op.ml]:
```suggestion
let _ = [::].:(-8)
```

(And rerun the test to confirm)

**Review by antalsz — CHANGES_REQUESTED** [review]
I went through and grepped the testsuite for `[:` and `[|`, and I'm satisfied that immutable array/comprehension copies of tests occur next to all existing array tests, which is the correctness condition for adding the new tests.  There's one requested change where we should use the immutable array indexing operator, but other than that the two questions for me are hereby addressed!

---

(... `|]`, `:]`.)

File: test/passing/tests/index_op.ml (hunk 1 above)
Comment (ccasin) [reply test/passing/tests/index_op.ml]: Thanks - fixed.

**Comment by ceastlund** [comment]
Pushed a newer rebase.

**Pull request closed without merging**